Add HRV feature extraction

hofaflo opened the pull request:
I started implementing basic heart rate variability (HRV) feature extraction. A few questions came up which I'd like to discuss here:

1. In our previous discussions we established that the RRI timeseries should be sampled regularly. In the few studies I've read now, an irregular RRI series is used for the time domain features and it is only resampled for the frequency domain features. Of course the number of RR intervals will differ for the time windows corresponding to each sleep stage, but using NaN-padded arrays keeps the calculations simple and loop-free. Do you see any issues with this approach?
1. Users will likely want to select which features should be extracted. To keep things uncluttered (and the code maintainable), we could maybe only allow the selection of feature groups (e.g. `hrv-timedomain`, `hrv-frequencydomain`, ...), but not of individual features. What do you think?
1. Should the returned feature matrix contain feature identifiers, e.g. by returning a DataFrame, or a dict of arrays? Or is it sufficient to document their order?
1. One commonly used time domain HRV feature is `pNN50`, which is (colloquially) the percentage of adjacent normal-to-normal RR intervals (i.e. NN intervals) differing by more than 50ms. [This report](https://doi.org/10.1161/01.CIR.93.5.1043), which defines a standard for HRV analysis, uses the total number of NN intervals as the denominator (instead of the number of pairs of NN intervals, which would be 1 less) - creating a biased estimate of the percentage (see also the discussion [here](https://github.com/Aura-healthcare/hrv-analysis/issues/24)). I'd prefer to deviate from the original literature here, since we'll usually analyze rather short RRI series (the standard was likely intended for long term ECG recordings).
1. Frequency domain features can't be calculated for time windows containing NaNs using only array operations. I see multiple options here:
    - iterate over the problematic arrays, slice them to contain no NaNs, piece the feature array together (this might not work for future features)
        - this could be limited to arrays containing only up to a specific percentage of NaNs
    - ignore NaNs where it is easily possible (like in the time domain features) and propagate them where it isn't
    - always propagate NaNs
1. Warnings posted with `warnings.warn` include the line the warning originated from, which makes the output confusing. There seems to be no (reliable) workaround, people suggest using logging instead - which would be nice to have for many parts of the package anyway - any objections?

Note:
- `doc/source/feature_extraction.md` includes descriptions of the features implemented so far. This should become a part of the documentation, though I've yet to figure out how to reference it from other parts of the docs.

hofaflo commented:
Re 1., each sleep stage annotation corresponds to (usually) a 30s time window. For a single time window, e.g. the average of all RR intervals within it can be calculated - creating a single feature for that time window. Other such features could be the standard deviation or maximum of the RR intervals within that time window. Of course not every 30s time window contains the same number of heartbeats (and thus RR intervals), so a 2d-array where each row contains the RR intervals of one 30s time window will have rows of different lengths. I see three approaches here:
- Use a list of 1d-arrays and iterate over them (-> this will be slow)
- Resample the irregular RRI timeseries to a regular one, so e.g. at 4Hz, each 30s time window would then have 120 entries (-> this might not be the most intuitive representation)
- Pad rows shorter than the longest one with NaNs, so the original representation of actual RR intervals is kept but calculations can be performed efficiently.

Re 2., okay, something to consider here: some features are based on a common step (e.g. the differences between RR intervals), others are based on other features. Allowing to select individual features could be implemented in different ways (considering code simplicity, I'd vote for the second option):
- Only calculate the selected features, which would require to either
  - perform a possible common step separately per feature (repeating identical calculations)
  - litter the code with if-statements
- Calculate all features of a group and select the requested ones afterwards. Depending on the feature selection, some calculations may be performed unnecessarily.

Re 3., 👍  There will be at least one categorical feature (sex), but likely nothing that can't be represented numerically, yes.

Re 4., 👍 `n-1` it is.

Re 5., NaNs can occur at signal borders, and (once that's implemented) in intervals where the RRI is nonsense. So in the best case  scenario (heartbeat detection yields plausible results over the whole night), the very first 30s time window will include NaNs for the time prior to the first detected heartbeat.

Re 6., Oh good to know, thanks for the link! :)

hofaflo commented:
Re 2., the new implementation allows selecting both entire feature groups (e.g. 'hrv-frequency') or single feature ids (e.g. 'meanHR'). If one feature of a group is selected, the entire group will be calculated, but `extract_features` will yield only the selected features.

Re 5., right, there is now a parameter to limit the maximum fraction of NaNs in an analysis window for which frequency domain windows will be calculated. For windows with a fraction of NaNs higher than `max_nans`, all frequency domain features will be `np.nan` 

`extract_features` now takes an iterable of `SleepRecord` objects and returns a list of feature matrices. Their dataclass definition will be introduced by #28, which is why the CI is failing right now. This makes extracting features for a single record a bit awkward (see [this example](https://github.com/hofaflo/sleepecg/blob/feature-extraction/examples/feature_extraction.py). Should we add the option to pass a single `SleepRecord`, in which case `extract_features` would return a single feature matrix instead of a list? This would make the function documentation slightly less clear.

hofaflo commented:
marked as draft because I realized a different API will be required to have the classification pipeline make sense

hofaflo commented:
Ready for review, @cbrnr :)

Some notes:
- I'll rebase this branch once #28 is merged, as this PR requires `SleepRecord`, which is introduced there.
- RRI preprocessing/validation is very rudimentary right now, I'll expand this in a future PR.

cbrnr commented:
@hofaflo I'm currently reviewing, and I was wondering if we could add a CI job which renders the documentation as a viewable artifact?

cbrnr commented:
In addition to my comment about doc rendering, just a quick heads up that `examples/feature_extraction.py` does not work for several reasons (`SleepRecord` import, wrong arguments when constructing a `SleepRecord` object, and the `extract_features` function giving an error because some attributes in `rec` are `None`).

It would be great if you could address these issues before I continue reviewing the rest.

cbrnr commented:
Third point: I think it would make sense to add all examples to our tests (preferably in a separate job). WDYT?

hofaflo commented:
ad CI for docs: great idea, I'll start right away!

ad example: fixed in 0598666 🙈 

ad examples as tests: I like the idea, but am unsure what you mean by "in a separate job" 🤔

hofaflo commented:
I've found the following possibilities for building docs in CI:
- [Sphinx Build Action](https://github.com/marketplace/actions/sphinx-build), with 2 ways to make the rendered docs available:
    - zip-download of the artifact
    - push to GitHub Pages
- an external CI provider, such as CircleCI (as used e.g. by mne-python and numpy), where you can click on "Details" next to the check to get to the docs
 
I'd prefer the latter option, as it would allow easy access of the docs while keeping the repo clean, what do you think?

cbrnr commented:
> ad example: fixed in 0598666 🙈

Great, thanks! I was wondering why `extract_features` returns the `stages` – it seems like this is just passed through the function without any modification. If this is the case, I think the function would not need to return it at all.

> ad examples as tests: I like the idea, but am unsure what you mean by "in a separate job" 🤔

I meant not adding it to an existing job such as `pre-commit` or `build_wheels`, but creating a separate dedicated `build-docs` job. (Side note: to be consistent I'd rename `build_wheels` and `build_sdist` to `build-wheels` and `build-sdist`).

> I've found the following possibilities for building docs in CI:

I'd also prefer keeping the repo clean, but if possible I'd stick with GitHub Actions. Using multiple providers is always more effort (to configure, to maintain, etc.). The sphinx-build docs mention that it can be paired with the upload-artifact action, which makes the built HTML available as an artifact. Would that work? If so, I'd prefer using that instead of CircleCI.

hofaflo commented:
> I was wondering why `extract_features` returns the `stages` – it seems like this is just passed through the function without any modification. If this is the case, I think the function would not need to return it at all.

Usually the parameter `records` will be a generator from one of the reader functions. If `extract_features` did not return the stages, the generator would have to be recreated outside of the function. Also, `extract_features` allows to resample the sleep stages to a new `sleep_stage_duration`, in which case they are modified.

> > ad examples as tests: I like the idea, but am unsure what you mean by "in a separate job" 🤔
>
> I meant not adding it to an existing job such as `pre-commit` or `build_wheels`, but creating a separate dedicated `build-docs` job.

I think we're mixing things up now :D So to be sure:
- For building the docs during a PR we'll have a new actions-workflow called `build-docs`
- We'll add the examples to the tests run via pytest to make sure they work with the current API (using a small dataset or single record to avoid them taking a long time)

> I'd also prefer keeping the repo clean, but if possible I'd stick with GitHub Actions. Using multiple providers is always more effort (to configure, to maintain, etc.). The sphinx-build docs mention that it can be paired with the upload-artifact action, which makes the built HTML available as an artifact. Would that work? If so, I'd prefer using that instead of CircleCI.

That would work, but the rendered docs would not be hosted somewhere. One would need to download the `artifact.zip`, extract and open the `index.html` manually. While not as nice, it'd probably be fine.

cbrnr commented:
> Usually the parameter `records` will be a generator from one of the reader functions. If `extract_features` did not return the stages, the generator would have to be recreated outside of the function. Also, `extract_features` allows to resample the sleep stages to a new `sleep_stage_duration`, in which case they are modified.

The second argument convinced me 😄.

> I think we're mixing things up now :D So to be sure:

I noticed that as well and edited my comment, but to be specific:

> - For building the docs during a PR we'll have a new actions-workflow called build-docs

Yes.

> - We'll add the examples to the tests run via pytest to make sure they work with the current API (using a small dataset or single record to avoid them taking a long time)

Well, I thought that we could also do that in a separate job. But then I looked at our jobs again, and it seems we're running `pytest` as part of the `build-wheels` job, right? Then of course we simply integrate the examples in the example folder.

But what do you mean by using a small dataset? The two examples are already self-contained and either use toy data or a specific record. I guess these should be fast enough to run?

> That would work, but the rendered docs would not be hosted somewhere. One would need to download the artifact.zip, extract and open the index.html manually. While not as nice, it'd probably be fine.

You're right, that's not very user-friendly and somehow defeats the purpose of quickly checking the rendered docs. Then I guess let's go for CircleCI then?

hofaflo commented:
> it seems we're running `pytest` as part of the `build-wheels` job, right? 

Exactly.

> But what do you mean by using a small dataset? The two examples are already self-contained and either use toy data or a specific record. I guess these should be fast enough to run?

I was thinking about the benchmark (which is also under `examples`), but we could just leave that one out :D

> You're right, that's not very user-friendly and somehow defeats the purpose of quickly checking the rendered docs. Then I guess let's go for CircleCI then?

Okay, I'll try to set it up in a different PR :)

cbrnr commented:
Yes, let's ignore the benchmark. Of course it would be great to be sure that this keeps working, but this would require adapting it to make it run faster. We can do that later, I'd just add the two examples for now.

Thanks for taking care of the CircleCI job!

cbrnr reviewed (commented):
Other than my minor comments LGTM! It would be nice to see some features "in action" though, e.g. a plot of all features over time with a real data set.

hofaflo commented on sleepecg/feature_extraction.py:
```diff
@@ -0,0 +1,555 @@
+    Calculate features from sleep data (e.g. heart rate).
+
+    Time and frequency domain heart rate variability (HRV) features are
+    calculated based on [1]_, [2]_ and [3]_. :ref:`feature_extraction`
```
[This](https://github.com/hofaflo/sleepecg/blob/feature-extraction/doc/source/feature_extraction.md) doc page

hofaflo commented:
> e.g. a plot of all features over time with a real data set

Good idea, I'd add this once we have a reader for a completely open dataset (e.g. [SLPDB](https://physionet.org/content/slpdb/1.0.0/)).

I added a parameter to `_hrv_frequencydomain_features` to avoid issuing warnings in situations where no feature from a problematic frequency range is requested and fixed all the other things :)

hofaflo commented:
The feature extraction test fails on 32bit Windows. It uses an analysis window of 3031 seconds to avoid warnings in `_hrv_frequencydomain_features`. This makes the intermediate result [here](https://github.com/hofaflo/sleepecg/blob/feature-extraction/sleepecg/feature_extraction.py#L341) (i.e. the return of `sliding_window_view`, before slicing) appear to be ~10.4GB large (which it actually isn't, since it's just a view). On a 32bit machine numpy is unhappy about that [here](https://github.com/numpy/numpy/blob/7125cdfa21081b91907a65b6c888791cf150a0b2/numpy/core/src/multiarray/ctors.c#L779-L785).

I think the simplest fix would be to use a much shorter window and just pass a list of `feature_ids` which won't trigger the warning.

cbrnr commented:
Alright, maybe I misunderstood why the warning is issued. It is kind of overkill to introduce a new parameter for the sole purpose of suppressing the warning. I thought that with the right amount of data we could solve the issue for the example.

Regarding the Win32 warning: why do we support Win32 in the first place?

If it's easier/quicker I'm happy to talk to you (e.g. in 30 minutes), LMK!

hofaflo commented:
@cbrnr ready :)

cbrnr commented:
Thanks @hofaflo!
